src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -546,33 +546,73 @@ private void doScheduledSearchForRound (RaptorState outputState) {
-                        // is evaluated first (providing updates to downstream stops), we need to check for earlier
-                        // trips that could be boarded at B.
-                        int bestTripIdx = onTrip;
-                        while (--bestTripIdx >= 0) {
-                            TripSchedule trip = pattern.tripSchedules.get(bestTripIdx);
-                            if (trip.headwaySeconds != null || !servicesActive.get(trip.serviceCode)) {
+
+                        // We have already boarded a trip on this pattern at an upstream stop; check for cases in which
+                        // it would be preferable to board the pattern at this stop instead.
+
+                        TripSchedule onTripSchedule = pattern.tripSchedules.get(onTrip);
+
+                        // First, check whether boarding the same trip at this stop instead of the upstream one would
+                        // be preferred. This will not affect total travel time, but it will affect the breakdown of
+                        // walk vs. wait. So if needed as an optimization, these blocks could be disabled unless
+                        // request.computePaths = true;
+                        boolean boardHereInstead = false;
+                        if (inputState.previous == null) {
+                            // In the first transit round, arriving at this stop before arriving at the upstream stop
+                            // implies shorter access time. And arriving earlier or at the same time implies more
+                            // buffer time to board the trip, so boarding at this stop is preferable to boarding at
+                            // the upstream stop when the non-strict inequality is satisfied.
+                            if (inputState.bestTimes[stop] <= inputState.bestTimes[boardStop]) {
+                                boardHereInstead = true;
+                            }
+                        } else {
+                            // In subsequent transit rounds, we want to compare the length of the transfer legs.
+                            // Boarding a trip at this stop is preferable if it requires a shorter transfer than
+                            // boarding the same trip at the upstream stop.
+                            int upstreamTransferFromStop = inputState.transferStop[boardStop];
+                            if (upstreamTransferFromStop != -1) {
+                                int upstreamTransferTime =
+                                        inputState.bestTimes[boardStop] - inputState.previous.bestTimes[upstreamTransferFromStop];
+                                int transferFromStop = inputState.transferStop[stop];
+                                int transferTime =
+                                        inputState.bestTimes[stop] - inputState.previous.bestTimes[transferFromStop];
+                                if (transferTime < upstreamTransferTime) {
```

Comment:
We're breaking the tie based on the one previous transfer time, but it seems roughly equivalent to me to maximize wait time over the whole path (as long as boarding time is the same). We are in fact already tracking cumulative waiting time in the array nonTransferWaitTime. Is there some way we could use that instead?

Reply:
To clarify: I'm not sure maximizing wait time is necessarily better than minimizing walk time. It seems like maximizing wait time might correspond better with an underlying reason for choosing the shorter path (all other things being equal it makes you less likely to miss your bus). But most importantly, we are already tracking nonTransferWaitTime cumulatively, and I was thinking it might result in exactly the same expression for the first round as for subsequent rounds. If it doesn't then there's no real advantage. Anyway since you factored that tie-breaker predicate out into a method, the presence of a conditional is less problematic because it no longer affects the readability of the logic one level up.